Gems/Terrain/Assets/Shaders/Terrain/TerrainClipmapGenerationPass.azsl

```diff
@@ -0,0 +1,142 @@
+        if ((macroMaterialData.m_flags & 1) == 0)
+        {
+            break; // No more macro materials for this tile
+        }
+
+        if (any(worldPosition < macroMaterialData.m_boundsMin.xy) || any(worldPosition > macroMaterialData.m_boundsMax.xy))
+        {
+            continue; // Macro material exists for this tile but is out of the bounds of this particular position
+        }
+
+        float2 macroUvSize = macroMaterialData.m_boundsMax - macroMaterialData.m_boundsMin;
+        macroUvSize.x = -1.0 / macroUvSize.x;
+        macroUvSize.y = 1.0 / macroUvSize.y;
+        float2 macroUv = (macroMaterialData.m_boundsMin - worldPosition) * macroUvSize;
+
+        // The macro uv gradient can vary massively over the quad because different pixels may choose different macro materials with different UVs.
+        // To fix, we use the world position scaled by the macro uv scale which should be fairly uniform across macro materials.
+        float2 ddx_macroUv = ddxPosition(clipmapLevel) * macroUvSize;
+        float2 ddy_macroUv = ddyPosition(clipmapLevel) * macroUvSize;
+
+        if (macroMaterialData.m_colorMapId != 0xFFFF)
+        {
+            macroColor = TerrainSrg::m_textures[macroMaterialData.m_colorMapId].SampleGrad(TerrainMaterialSrg::m_sampler, macroUv, ddx_macroUv, ddy_macroUv).rgb;
+            macroColor = TransformColor(macroColor, ColorSpaceId::LinearSRGB, ColorSpaceId::ACEScg);
+        }
+        else
+        {
+            macroColor = TerrainMaterialSrg::m_baseColor.rgb;
+        }
+
+        if (macroMaterialData.m_normalMapId != 0xFFFF)
+        {
+            bool flipX = macroMaterialData.m_flags & 2;
+            bool flipY = macroMaterialData.m_flags & 4;
+            float2 sampledValue = TerrainSrg::m_textures[macroMaterialData.m_normalMapId][macroUv].xy;
+            if(flipX)
+            {
+                sampledValue.y = -sampledValue.y;
+            }
+            if(flipY)
+            {
+                sampledValue.x = -sampledValue.x;
+            }
+            macroNormal = normalize(GetTangentSpaceNormal_Unnormalized(sampledValue, macroMaterialData.m_normalFactor));
+        }
+        else
+        {
+            macroNormal = float3(0.0, 0.0, 1.0);
+        }
+
+        break;
+    }
+}
+
+[numthreads(32,32,1)]
+void MainCS(
+    uint3 dispatchThreadID : SV_DispatchThreadID)
+{
+    uint2 pixelPosition = dispatchThreadID.xy;
+
+    for (uint clipmapLevel = 0; clipmapLevel < ClipmapStackSize; ++clipmapLevel)
+    {
+        float2 worldPosition = GetCurrentWorldPosition(pixelPosition, clipmapLevel);
+        if (any(worldPosition < TerrainSrg::m_clipmapData.m_worldBounds.xy) || any(worldPosition > TerrainSrg::m_clipmapData.m_worldBounds.zw))
+        {
+            PassSrg::m_macroColorClipmaps[uint3(dispatchThreadID.xy, clipmapLevel)] = float4(TerrainMaterialSrg::m_baseColor.rgb, 1.0);
```

Comment:
Nit: suggest pulling out `uint3 texelAddress = uint3(dispatchThreadID.xy, clipmapLevel);` (or some other variable name) since this is used a lot in this file.